File: 백준/Silver/2839. 설탕 배달/설탕 배달.py
 #다이나믹 프로그래밍
 #2839번 , 설탕 배달
 
n=int(input())
dp=[5001]*(n+5)

dp[3]=1
dp[5]=1

for i in range(6,n+1):
   dp[i]=min(dp[i-3],dp[i-5])+1

if dp[n]<5001:
    print(dp[n])
else:
    print(-1)

File: 백준/Silver/2839. 설탕 배달/설탕 배달.cc
#include <iostream>
#include <vector>

using namespace std;

vector<int> dp;

int solve(int n) {
	if (n < 0) return 5001;
	if (n == 0) return 0;
	if (dp[n] == -1) {
		dp[n] = min(solve(n - 3), solve(n - 5)) + 1;
	}
	return dp[n];
}
int main() {
	ios::sync_with_stdio(false);
	cin.tie(0); cout.tie(0);

	int n; cin >> n;
	dp.assign(n + 1, -1);

	if (solve(n) >= 5001) cout << "-1";
	else cout << solve(n);
	return 0;
}